mailbox/tika/src/main/java/org/apache/james/mailbox/tika/CachingTextExtractor.java

```diff
@@ -0,0 +1,131 @@
+                                MetricFactory metricFactory, GaugeRegistry gaugeRegistry) {
+        this.underlying = underlying;
+        this.weightMetric = metricFactory.generate("textExtractor.cache.weight");
+
+        Weigher<String, ParsedContent> weigher =
+            (key, parsedContent) -> {
+                int size = getSize(parsedContent);
+                weightMetric.add(size);
+                return size;
+            };
+        RemovalListener<String, ParsedContent> removalListener =
+            notification -> Optional.ofNullable(notification.getValue())
+                .map(this::getSize)
+                .ifPresent(weightMetric::remove);
+
+        this.cache = CacheBuilder.<String, String>newBuilder()
+            .expireAfterAccess(cacheEvictionPeriod.toMillis(), TimeUnit.MILLISECONDS)
+            .maximumWeight(cacheWeightInBytes)
+            .weigher(weigher)
+            .recordStats()
+            .removalListener(removalListener)
+            .build();
+        recordStats(gaugeRegistry);
+    }
+
+    public void recordStats(GaugeRegistry gaugeRegistry) {
+        gaugeRegistry
+            .register(
+                "textExtractor.cache.hit.rate",
+                () -> cache.stats().hitRate())
+            .register(
+                "textExtractor.cache.hit.count",
+                () -> cache.stats().hitCount());
+            gaugeRegistry.register(
+                "textExtractor.cache.load.count",
+                () -> cache.stats().loadCount())
+            .register(
+                "textExtractor.cache.eviction.count",
+                () -> cache.stats().evictionCount())
+            .register(
+                "textExtractor.cache.load.exception.rate",
+                () -> cache.stats().loadExceptionRate())
+            .register(
+                "textExtractor.cache.load.miss.rate",
+                () -> cache.stats().missRate())
+            .register(
+                "textExtractor.cache.load.miss.count",
+                () -> cache.stats().missCount())
+            .register(
+            "textExtractor.cache.size",
+            cache::size);
+    }
+
+    private int getSize(ParsedContent parsedContent) {
+        return parsedContent.getTextualContent()
+            .map(String::length)
+            .map(this::utf16LengthToBytesCount)
+            .orElse(0);
+    }
+
+    private int utf16LengthToBytesCount(Integer value) {
+        return value * 2;
+    }
+
+    @Override
+    public ParsedContent extractContent(InputStream inputStream, String contentType) throws Exception {
+        byte[] bytes = IOUtils.toByteArray(inputStream);
+        String key = DigestUtils.sha256Hex(bytes);
+
+        ParsedContent cachedValue = cache.getIfPresent(key);
+        if (cachedValue != null) {
+            return cachedValue;
+        }
+
+        ParsedContent realValue = underlying.extractContent(new ByteArrayInputStream(bytes), contentType);
+        cache.put(key, realValue);
```

Comment:
I may missed something but I would have think you need to provide a CacheLoader to guava to prevent race conditions

Reply:
I gonna revert that change then ;-)

Reply:
I don't see how this report to Matthieu's comment

Reply:
Simple:

Duc code optimization proposed here : https://github.com/linagora/james-project/pull/1434#discussion_r191351151 in order to simplify error handling that I contributed here: https://github.com/linagora/james-project/pull/1434/commits/bda4e0c7d78ee4585c07e5e7c03625c45e7475e6 broke the interesting feature Guava cahes have regarding concurrency (as tested here https://github.com/linagora/james-project/pull/1434/commits/844778742a21e9a2277d05fc1e2d9850283a2f54 and reported by matthieu here https://github.com/linagora/james-project/pull/1434#discussion_r191372051).

Given that new knowledge, it's clear https://github.com/linagora/james-project/pull/1434#discussion_r191351151 is not desirable.